lib/CL/devices/remote/network_discovery.c

```diff
@@ -0,0 +1,687 @@
+      char *domains, *token, *saveptr = NULL;
+      domains = strdup (env);
+      /* .local + at least one other domain in env as env is not null */
+      domain_count = 1 + 1;
+      token = strtok_r (domains, ",", &saveptr);
+      while (token)
+        {
+          token = strtok_r (NULL, ",", &saveptr);
+          domain_count++;
+        }
+      token = NULL;
+      saveptr = NULL;
+      POCL_MEM_FREE (domains);
+
+      /* Allocate memory for Avahi browser for each domain */
+      service_browser = (AvahiServiceBrowser **)malloc (
+        domain_count * sizeof (AvahiServiceBrowser *));
+
+      /* Initialize browser for .local domain */
+      service_browser[0] = avahi_service_browser_new (
+        avahi_client, AVAHI_IF_UNSPEC, AVAHI_PROTO_UNSPEC,
+        POCL_REMOTE_DNS_SRV_TYPE_ENV, NULL, 0, avahi_browser_callback, NULL);
+      POCL_GOTO_ERROR_ON (
+        !service_browser[0], CL_OUT_OF_RESOURCES,
+        "Avahi failed to create service browser with error: %s\n",
+        avahi_strerror (avahi_client_errno (avahi_client)));
+
+      /* Loop to setup browsers in the env specified domains */
+      domains = strdup (env);
+      token = strtok_r (domains, ",", &saveptr);
+      while (token)
+        {
+          service_browser[0] = avahi_service_browser_new (
+            avahi_client, AVAHI_IF_UNSPEC, AVAHI_PROTO_UNSPEC, token, NULL, 0,
+            avahi_browser_callback, NULL);
+          POCL_GOTO_ERROR_ON (
+            !service_browser[0], CL_OUT_OF_RESOURCES,
+            "Avahi failed to create service browser with error: %s\n",
+            avahi_strerror (avahi_client_errno (avahi_client)));
+
+          token = strtok_r (NULL, ",", &saveptr);
+        }
+
+      POCL_MEM_FREE (domains);
+    }
+  else
+    {
+      /* Browser only for .local if env for domains is null */
+      domain_count = 1;
+      service_browser
+        = (AvahiServiceBrowser **)malloc (sizeof (AvahiServiceBrowser *));
+
+      service_browser[0] = avahi_service_browser_new (
+        avahi_client, AVAHI_IF_UNSPEC, AVAHI_PROTO_UNSPEC,
+        POCL_REMOTE_DNS_SRV_TYPE_ENV, NULL, 0, avahi_browser_callback, NULL);
+      POCL_GOTO_ERROR_ON (
+        !service_browser[0], CL_OUT_OF_RESOURCES,
+        "Avahi failed to create service browser with error: %s\n",
+        avahi_strerror (avahi_client_errno (avahi_client)));
+    }
+
+  /* Start the avahi thread*/
+  avahi_errcode = avahi_threaded_poll_start (avahi_threaded_poll);
+  POCL_GOTO_ERROR_ON ((avahi_errcode < 0), CL_OUT_OF_RESOURCES,
+                      "Avahi failed to start threaded poll with error: %s\n",
+                      avahi_strerror (avahi_errcode));
+  POCL_MSG_PRINT_REMOTE ("Avahi Browser: Browsing started \n");
+
+  return errcode;
+
+ERROR:
+
+  clear_avahi ();
+  return errcode;
+}
+#endif
+
+/*****************************************************************************/
+#if defined(ENABLE_REMOTE_DISCOVERY_DHT)
+
+const char *common_key = NULL;
```

Comment:
Should this really be a global variable?

Reply:
Only for this file, therefore also static.